core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -183,6 +185,11 @@ class KafkaApis(val requestChannel: RequestChannel,
     val updateMetadataResponse =
       if (authorize(request.session, ClusterAction, Resource.ClusterResource)) {
         replicaManager.maybeUpdateMetadataCache(correlationId, updateMetadataRequest, metadataCache)
+        if (controller.isActive && !updateMetadataRequest.partitionStates.isEmpty) {
```

Comment:
The controller only changes when it has soft failure, typically due to GC. It should be rare if the GC is configured properly.